src/client/line.cpp

```diff
@@ -0,0 +1,556 @@
+	axis_rot_mat.setRotationAxisRadians(params.axis_angle, dir);
+	trans_mat.setTranslation(pos.first);
+
+	res_mat = trans_mat * axis_rot_mat * rot_mat;
+
+	res_mat.transformVect(positions[0]);
+	res_mat.transformVect(positions[1]);
+	res_mat.transformVect(positions[2]);
+	res_mat.transformVect(positions[3]);
+
+	core::plane3df line_plane(positions[1], positions[2], positions[3]);
+
+	video::ITexture *tex = mat.getTexture(0);
+
+	v2f ll_corner_coords(0.0f, 0.0f);
+	v2f ur_corner_coords(1.0f, 1.0f);
+
+	if (tex) {
+		core::dimension2du tex_size = tex->getSize();
+
+		if (params.animation.type != TAT_NONE) {
+			v2u32 frame_size;
+			params.animation.determineParams(tex_size, nullptr, nullptr, &frame_size);
+
+			f32 aspect = (f32)frame_size.Y / frame_size.X;
+			f32 tcoords_u = dist * aspect / params.width;
+
+
+			ll_corner_coords = params.animation.getTextureCoords(tex_size, current_animation_frame);
+			ur_corner_coords.X = ll_corner_coords.X + (f32)frame_size.X / tex_size.Width * tcoords_u;
+			ur_corner_coords.Y = ll_corner_coords.Y + (f32)frame_size.Y / tex_size.Height;
+		}
+		else {
+			f32 aspect = (f32)tex_size.Height / tex_size.Width;
+			f32 tcoords_u = dist * aspect / params.width;
+
+			ur_corner_coords.X = tcoords_u;
+		}
+	}
+	video::S3DVertex vertices[4] = {
+		video::S3DVertex(positions[0], line_plane.Normal, color.first, ll_corner_coords),
+		video::S3DVertex(positions[1], line_plane.Normal, color.first, v2f(ll_corner_coords.X, ur_corner_coords.Y)),
+		video::S3DVertex(positions[2], line_plane.Normal, color.second, ur_corner_coords),
+		video::S3DVertex(positions[3], line_plane.Normal, color.second, v2f(ur_corner_coords.X, ll_corner_coords.Y))
+	};
+
+	video::IVideoDriver *vdrv = SceneManager->getVideoDriver();
+
+	vdrv->setTransform(video::ETS_WORLD, core::IdentityMatrix);
+	vdrv->setMaterial(mat);
+
+	u16 indices[] = {0, 1, 2, 2, 3, 0};
+
+	vdrv->drawIndexedTriangleList(vertices, 4, indices, 2);
+}
+
+void LineSceneNode::calc3DCircle(bool first, video::S3DVertex *verts, u16 *inds, u32 verts_c) const
+{
+	const v3f &center_pos = first ? pos.first : pos.second;
+	v3f normal(0.0f, 0.0f, first ? -1.0f : 1.0f);
+	video::SColor col = first ? color.first : color.second;
+
+	// Code of the future animation support for this type
+	/*v2f ll_corner_coords(0.0f, 0.0f);
+	v2f ur_corner_coords(1.0f, 1.0f);
+
+	if (params.animation.type != TAT_NONE)
+	{
+		core::dimension2du tex_size = mat.getTexture(0)->getSize();
+
+		v2u32 frame_size;
+		params.animation.determineParams(tex_size, nullptr, nullptr, &frame_size);
+
+		ll_corner_coords = params.animation.getTextureCoords(tex_size, current_animation_frame);
+		ur_corner_coords = v2f(
+			ll_corner_coords.X + (f32)frame_size.X / tex_size.Width,
+			ll_corner_coords.Y + (f32)frame_size.Y / tex_size.Height
+		);
+	}
+
+	v2f center_tcoords = (ll_corner_coords + ur_corner_coords) / 2.0f;
+	v2f shift(ll_corner_coords.X - center_tcoords.X, 0.0f);*/
```

Comment:
Please put it elsewhere. This can be put on any paste service (e.g. GitHub Gist) and put its link in this PR.